control-plane/api-gateway/cache/consul.go

```diff
@@ -278,6 +307,66 @@ func (c *Cache) Write(ctx context.Context, entry api.ConfigEntry) error {
 	return nil
 }
 
+func (c *Cache) ensurePolicy(client *api.Client) (string, error) {
+	policy := c.gatewayPolicy()
+
+	created, _, err := client.ACL().PolicyCreate(&policy, &api.WriteOptions{})
+
+	if isPolicyExistsErr(err, policy.Name) {
+		existing, _, err := client.ACL().PolicyReadByName(policy.Name, &api.QueryOptions{})
+		if err != nil {
+			return "", err
+		}
+		return existing.ID, nil
+	}
+	if err != nil {
+		return "", err
+	}
+	return created.ID, nil
+}
+
+func (c *Cache) ensureRole(client *api.Client) (string, error) {
+	policyID, err := c.ensurePolicy(client)
+	if err != nil {
+		return "", err
+	}
+
+	aclRoleName := "managed-gateway-acl-role"
+	aclRole, _, err := client.ACL().RoleReadByName(aclRoleName, &api.QueryOptions{})
+	if err != nil {
+		return "", err
+	}
+	if aclRole != nil {
+		return aclRoleName, nil
+	}
+
+	role := &api.ACLRole{
+		Name:        aclRoleName,
+		Description: "ACL Role for Managed API Gateways",
+		Policies:    []*api.ACLLink{{ID: policyID}},
+	}
+
+	_, _, err = client.ACL().RoleCreate(role, &api.WriteOptions{})
+	return aclRoleName, err
+}
+
+func (c *Cache) gatewayPolicy() api.ACLPolicy {
+	var data bytes.Buffer
+	if err := gatewayTpl.Execute(&data, templateArgs{
```

Comment:
this won't change during runtime right? is there a way we could build this once at start and just it use through the life of the program?

Reply:
We could wrap it in a `sync.Once` if we really want to, but part of it is that the namespace bit depends on the runtime flags you pass this. The template itself is compiled once during an init phase.